### PR TITLE
Apply cross-origin headers to all incoming requests.

### LINKED ISSUE #363: Allow OPTIONS method against token & collection routes
When CORS is enabled (config:`"cors": true`) we send back `'Access-Control-Allow-Methods', 'GET, POST, PUT, DELETE, OPTIONS'`, but then `OPTIONS` requests to /token and any collection route 404

### PR DESCRIPTION
**Pull request by phyllisstein**
### Does this resolve an issue?
Fix/Close #363 

### Description of changes proposed in this pull request
Creates a global cross-origin middleware that applies spec-compliant CORS headers to all incoming requests.

### Who should review this pull request?
@jimlambie, @adamkdean 

### Testing
Integration tests are passing---there's no way I can think of to do any meaningful acceptance testing, but I have confidence that this is a valid tack.

### CODE REVIEW

**Comment by jimlambie** [comment]
@phyllisstein this looks great - we'll review/merge/release tomorrow morning. The inclusion of tests much appreciated!

**Comment by jimlambie** [comment]
@eduardoboucas we'll need to port this to the 2.2.x branch, I believe

File: dadi/lib/cors/index.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+const config = require(path.join(__dirname, '/../../../config.js'))
+const vary = require('vary')
+
+module.exports = function cors (server) {
+  server.app.use((req, res, next) => {
+    if (config.get('cors') !== true) {
+      return next()
+    }
+
+    const method = req.method && req.method.toUpperCase && req.method.toUpperCase()
```
Comment (eduardoboucas) [inline dadi/lib/cors/index.js]: Is it necessary to check for the existence of both `req.method` and `req.method.toUpperCase` before using it? Would `req.method` ever be anything other than a string?

Comment (jimlambie) [reply dadi/lib/cors/index.js]: also, we lowercase it everywhere else ;)

Comment (phyllisstein) [reply dadi/lib/cors/index.js]: 😃Fair enough! Just pushed a less conservative check.

**Comment by phyllisstein** [comment]
Thanks for the feedback! Should I change the base branch to `2.2.x` and see if Git can manage that, rather than porting it by hand?

**Comment by jimlambie** [comment]
@phyllisstein yes please, that would be much appreciated!

**Comment by jimlambie** [comment]
I've made a mess of this. Closing this PR - @phyllisstein please reopen against `develop`, I'm doing a new one against 2.2.x

thanks!

**Comment by phyllisstein** [comment]
Done: https://github.com/dadi/api/pull/376! Sorry for the not-so-hot suggestion about swapping branches, that never seems to work the way I expect.

**Pull request closed without merging**